util/ranger/detacher.go

```diff
@@ -133,61 +136,189 @@ func getEqOrInColOffset(expr expression.Expression, cols []*expression.Column) i
+	conditions = removeAccessConditions(conditions, accessConds)
+	eqOrInCount := len(accessConds)
+	if eqOrInCount == len(cols) {
+		// If curIndex equals to len of index columns, it means the rest conditions haven't been appended to filter conditions.
+		filterConds = append(filterConds, conditions...)
+		ranges, err = buildCNFIndexRange(sc, cols, tpSlice, lengths, eqOrInCount, accessConds)
+		if err != nil {
+			return nil, nil, nil, 0, errors.Trace(err)
 		}
+		return ranges, accessConds, filterConds, eqCount, nil
 	}
-	return accessConds, filterConds
+	accessConds, filterConds, err = extractAccessAndFilterConds(conditions, accessConds, filterConds, cols[eqOrInCount].ColName, lengths[eqOrInCount])
+	ranges, err = buildCNFIndexRange(sc, cols, tpSlice, lengths, eqOrInCount, accessConds)
+	return ranges, accessConds, filterConds, eqCount, errors.Trace(err)
 }
 
-// DetachIndexConditions will detach the index filters from table filters.
-// It will first find the point query column and then extract the range query column.
-func DetachIndexConditions(conditions []expression.Expression, cols []*expression.Column,
-	lengths []int) (accessConds []expression.Expression, filterConds []expression.Expression) {
-	accessConds = make([]expression.Expression, len(cols))
-	var equalOrInCount int
+func extractEqAndInCondAndBuildRange(conditions []expression.Expression, cols []*expression.Column,
+	lengths []int) (accesses, filters []expression.Expression, err error) {
+	accesses = make([]expression.Expression, len(cols))
 	for _, cond := range conditions {
 		offset := getEqOrInColOffset(cond, cols)
 		if offset != -1 {
-			accessConds[offset] = cond
+			accesses[offset] = cond
 		}
 	}
-	for i, cond := range accessConds {
+	for i, cond := range accesses {
 		if cond == nil {
-			accessConds = accessConds[:i]
-			equalOrInCount = i
+			accesses = accesses[:i]
 			break
 		}
 		if lengths[i] != types.UnspecifiedLength {
-			filterConds = append(filterConds, cond)
+			filters = append(filters, cond)
+		}
+	}
+	return accesses, filters, nil
+}
+
+// detachDNFCondAndBuildRangeForIndex will detach the index filters from table filters when it's a DNF.
+// We will detach the conditions of every DNF items, then compose them to a DNF.
+func detachDNFCondAndBuildRangeForIndex(sc *stmtctx.StatementContext, condition *expression.ScalarFunction,
+	cols []*expression.Column, newTpSlice []*types.FieldType, lengths []int) ([]*NewRange, []expression.Expression, bool, error) {
+	firstColumnChecker := &conditionChecker{
+		colName:       cols[0].ColName,
+		shouldReserve: lengths[0] != types.UnspecifiedLength,
+		length:        lengths[0],
+	}
+	rb := builder{sc: sc}
+	dnfItems := expression.FlattenDNFConditions(condition)
+	newAccessItems := make([]expression.Expression, 0, len(dnfItems))
+	var totalRanges []*NewRange
+	hasResidual := false
+	for _, item := range dnfItems {
+		if sf, ok := item.(*expression.ScalarFunction); ok && sf.FuncName.L == ast.LogicAnd {
```

Comment:
There are many similar checks here. Maybe we can write two functions called `isCNF(expr)` and `isDNF(expr)`

Reply:
So some code like this?
```go
if isCNF(expr) {
    sf := expr.(*expression.ScalarFunction);
    doSomeThing...;
}
```

Reply:
how about `sf := tryToGetCNF` and `if sf == nil; continue`

Reply:
i think we don't need to wrap a type assertion to a function.